text/0017-gas-futures.md

```diff
@@ -0,0 +1,144 @@
+
+**Example:** Holders of the token `BST-2020-09-01-01000` will be able to include transactions in the 1000th block that is proposed on September 1, 2020 at the price they have agreed on at the auction. The auction takes place exactly 180 days before the delivery date, for the whole day, on March 5, 2020.
+
+### Futures size
+
+The protocol parameter `BST_SIZE` determines how much gas a BST corresponds to. In Casper, currently 1e12 block gas limit is envisioned. We take one-thousandth of a block to be a sufficiently small size for a futures contract. Therefore, we take `BST_SIZE = 1_000_000_000`. Each BST will grant the holder the right to include 1 billion gas worth of transactions.
+
+The maximum supply of a BST token for a given block is found by `MAX_FUTURES_GAS_LIMIT / BST_SIZE`.
+
+### Futures gas limit
+
+`FUTURES_GAS_LIMIT` varies for each block, depending on the amount of BST sold off for that block. If no BST were sold, it is simply equal to zero. We dictate `BST_SOLD[i] <= MAX_FUTURES_GAS_LIMIT / BST_SIZE` for any block `i`. Then `FUTURES_GAS_LIMIT` for `i` is determined by
+
+```python
+FUTURES_GAS_LIMIT[i] = BST_SOLD[i] * BST_SIZE
+```
+
+<!-- ### Futures quota -->
+
+<!-- `FUTURES_GAS_LIMIT` varies for each day, depending on the amount of BST sold off for that day. If no BST were sold, it is simply equal to zero. There is an upper limit to the amount of BST that can be sold at any given day, called `FUTURES_QUOTA`. We dictate to condition `FUTURES_SOLD[i] <= FUTURES_QUOTA` for any day `i`. Then `FUTURES_GAS_LIMIT` for `i` is determined by -->
+
+<!-- ```python -->
+<!-- FUTURES_GAS_LIMIT[i] = BLOCK_GAS_LIMIT * FUTURES_SOLD[i] * BST_SIZE / GAS_PER_DAY -->
+<!-- ``` -->
+
+<!-- **Example:** To determine `FUTURES_QUOTA`, we work our way back from the `FUTURES_GAS_LIMIT` we desire. Let us assume that we want 20% of block space to be reserved for futures transactions. This sets our upper limit in the case which the whole quota is sold off, i.e. `FUTURES_SOLD[i] = FUTURES_QUOTA`. Solving the equation with these values, we obtain `FUTURES_QUOTA = 0.2 * GAS_PER_DAY / BST_SIZE`. Assuming `2**14 = 16384 ms` rounds, we have `60*60*24/16.384 ~ 5273` blocks per day, giving us `GAS_PER_DAY = 5.273e15`. Finally, we compute `FUTURES_QUOTA = 0.2 * 5.273e15 / 1e9 = 1054600` BST per day. -->
+
+### Auction
+
+The auction for `GAS-YYYY-MM-DD-NNNNN` begins at tick `X - GAS_AUCTION_PERIOD * TICKS_IN_DAY` where `X` is the tick corresponding to `YYYY-MM-DD` 00:00 in UTC, and `TICKS_IN_DAY = 24*60*60*1000` is the number of ticks in one day. The auction can go on until the whole quota is sold off.
+
+The auction price is determined by the validators collectively, who announce the price they are willing to accept. The futures price is computed by taking a weighted average:
+
+```python
+auction_price = sum(weight[v] * price[i] for v in validators) / sum(weight[v] for v in validators)
```

Comment:
As discussed, this should probably be a (weighted) median instead of average. But that's probably a question for a separate discussion.